servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Completable.java

```diff
@@ -1387,7 +1388,24 @@ public final Completable liftAsync(CompletableOperator operator) {
      * {@link Subscriber}.
      */
     public final Completable publishOn(Executor executor) {
-        return PublishAndSubscribeOnCompletables.publishOn(this, executor);
+        return PublishAndSubscribeOnCompletables.publishOn(this, () -> Boolean.TRUE::booleanValue, executor);
+    }
+
+    /**
+     * Creates a new {@link Completable} that map use the passed {@link Executor} to invoke {@link Subscriber}
+     * methods.
+     * This method does <strong>not</strong> override preceding {@link Executor}s, if any, specified for {@code this}
+     * {@link Completable}. Only subsequent operations, if any, added in this execution chain will use this
+     * {@link Executor}.
+     *
+     * @param executor {@link Executor} to use.
+     * @param shouldOffload provides a hint whether offloading to executor can be omitted. Offloading may still occur
+     * even if {@code false} is returned in order to preserve signal ordering.
+     * @return A new {@link Completable} that will use the passed {@link Executor} to invoke all methods on the
+     * {@link Subscriber}.
+     */
+    public final Completable publishOn(Executor executor, Supplier<? extends BooleanSupplier> shouldOffload) {
```

Comment:
1. Based on the test that utilizes this method, it looks like the `Supplier` is not required. Users who want to have a `BooleanSupplier` per subscribe, can wrap this operator with `defer`. Can we simplify API and remove the `Supplier`?
2. `BooleanSupplier` doesn't give users any information about which method is invoked. Did you consider  `Predicate` instead that takes an enum value saying what method is invoked? Examples: users may want to offload only terminal signals for the `Publisher` or they may want to offload `onComplete()` but not `onError`.

Reply:
> * Based on the test that utilizes this method, it looks like the `Supplier` is not required. Users who want to have a `BooleanSupplier` per subscribe, can wrap this operator with `defer`. Can we simplify API and remove the `Supplier`?

I like this idea as it is a nice simplification.

> * `BooleanSupplier` doesn't give users any information about which method is invoked. Did you consider  `Predicate` instead that takes an enum value saying what method is invoked? Examples: users may want to offload only terminal signals for the `Publisher` or they may want to offload `onComplete()` but not `onError`.

The intention was not to provide such fine-grained decisions but rather to base the decision to offload on the context such as "currently on the io thread". There doesn't yet seem to be a clear case where this is necessary but it could be added if we find a use case.

Reply:
@Scottmitch WDYT about removing the extra layer of `Supplier` since the few who might want a `BooleanSupplier` per subscribe can use defer?

Reply:
+1, we may also revisit `Supplier` for `scanWith` and `scanWithLifetime`, those can also use `defer`. Non of the other operators use a `Supplier` to provide a function (map/flatMap/etc) per subscribe. So, better to be consistent.

Reply:
`Supplier` was removed in #1723, thank you @bondolo!